### PR TITLE
Xhr2 and node

### PR DESCRIPTION
**Pull request by latentflip**
Extending #45 (merge that one first) to test node support, by switching xhr for request as per #21 

Currently some tests are failing:

* ensuring that `"application/json"` is set in the headers
* ensuring `beforeSend` is called

I haven't dug into why yet.

### CODE REVIEW

File: core.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      }
+
+      // Turn a jQuery.ajax formatted request into xhr compatible
+      params.method = params.type;
+
+      var ajaxSettings = assign(params, options);
+
+      // Make the request. The callback executes functions that are compatible
+      // With jQuery.ajax's syntax.
+      var request = options.xhr = options.xhrImplementation(ajaxSettings, function (err, resp, body) {
```
Comment (naugtur) [inline core.js]: I never understood how overwriting options.xhr with itself benefited this code, but now it's probably going to be harmful, as `request` doesn't return itself.

File: core.js [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+          params.beforeSend = function (req) {
+              for (var key in ajaxConfig.xhrFields) {
+                  req[key] = ajaxConfig.xhrFields[key];
+              }
+              if (beforeSend) return beforeSend.apply(this, arguments);
+          };
+          params.xhrFields = ajaxConfig.xhrFields;
+      } else {
+          params.beforeSend = ajaxConfig.beforeSend;
+      }
```
Comment (naugtur) [inline core.js]: both `xhrFields` and `beforeSend` don't seem to make sense in nodejs. `request` won't handle anything other than can be done by passing options. Ok, there's one thing - streams. But I don't think anyone wants to use ampersand-sync with binary data.

File: package.json [hunk 3]
```diff
@@ -44,13 +45,14 @@
   ],
   "license": "MIT",
   "main": "ampersand-sync.js",
+  "browser": "ampersand-sync-browser.js",
```
Comment (latentflip) [inline package.json]: I don't believe that format works with webpack by default, that was my main concern with this approach.

**Comment by naugtur** [comment]
I also looked into the failing tests with `application/json` header and saw that the test worked only because `xhr` reused the object passed in as headers. If a copy of that object is used, the reference will not point to the object that actually holds the expected header.

IMHO, unit tests need to replace the request maker with something that would handle assertions and integration tests should ensure that sync works with both xhr and request.

If you agree with the above, I could get it done on Saturday (if I wake up before dinner time ;) )

File: package.json (hunk 3 above)
Comment (naugtur) [inline package.json]: I didn't know that. So there's no other option for good defaults.

**Comment by HenrikJoreteg** [comment]
btw, @naugtur, really appreciate all your work on this. I've added you as a member of the community team so you can easily contribute to this PR directly. If you're ok with it, I would happily denote that on this page as well: http://ampersandjs.com/contribute

**Comment by naugtur** [comment]
That's cool, thanks! 
I feel like I should do a bit more and wait for 4.0 to get released before my name shows up at http://ampersandjs.com/contribute, but I'm happy to join.

It's getting late. I'll get back to this PR on Friday and I'll rebuild the tests to be independent of minor implementation details in xhr.

File: core.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
+/*$AMPERSAND_VERSION*/
+module.exports = function (xhr) {
+  var result = require('lodash.result');
```
Comment (naugtur) [inline core.js]: Is there a reason for those requires to be inside the exported function?
Unless there are circular dependencies anywhere, there should be no benefit from requiring from within a function that is called later.

Comment (latentflip) [inline core.js]: No, just me not thinking straight :)

Philip Roberts
&yet

File: core.js [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  return function (method, model, options) {
+      var type = methodMap[method];
+      var headers = {};
+
+      // Default options, unless specified.
+      defaults(options || (options = {}), {
+          emulateHTTP: false,
+          emulateJSON: false,
+          // overrideable primarily to enable testing
+          xhrImplementation: xhr
```
Comment (naugtur) [inline core.js]: I looked for documentation mentioning this option, but it seems undocumented. For the sake of being more cross-platform we might consider changing it to one of these:
requestOverride
requestEngine
httpRequestImplementation

File: test/integration.js [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            t.pass('received an expected error');
+            t.end();
+        }
+    });
+});
+
+
+
+if (ImInBrowser) {
+    test('should call user provided beforeSend function. (when in the browser)', function (t) {
```
Comment (naugtur) [inline test/integration.js]: beforeSend will only be tested in the browser

File: test/integration.js [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        });
+        var m = new Me();
+        var requestStream = m.fetch({});
+        var writable = concatStream({
+            encoding: 'string'
+        }, function (data) {
+            t.equal(JSON.parse(data).foo, 'bar');
+            t.end();
+        });
+        requestStream.pipe(writable);
```
Comment (naugtur) [inline test/integration.js]: This is one of the `request`'s main features, but the test is here only to check if stream support is exposed, I don't have a reasonable use-case in mind right now.

File: test/unit.js [hunk 8]
```diff
@@ -0,0 +1,235 @@
+var test = require('tape');
+var syncCore = require('../core');
+var reqStub = require('./helpers/requestStub');
+var modelStub = require('./helpers/modelStub');
+var Model = require('ampersand-model');
+
+var sync = syncCore(reqStub());
```
Comment (naugtur) [inline test/unit.js]: Plug the stub in once, it exposes options with which it's been called each time.

**Comment by naugtur** [comment]
That's it for decoupling the unit tests from 3rd party code.
And I didn't have to do anything to core to make them pass (except for beforeSend, see above)

And btw, I noticed that when initially splitting the tests into unit and integration I hit the autoformatter and re-formated everything with my 4space setting. My bad. Is it worth the mess in commit history to make it consistent again? 

Also, ampersand-model uses 4spaces, I'm confused. ;)

**Comment by latentflip** [comment]
This is all looking good to me, I'm not sure if we've decided on 2 or 4 space across the board yet :/

@HenrikJoreteg et al?

**Comment by HenrikJoreteg** [comment]
Currently we're on 4 spaces on all ampersand modules, we should stick to that. (should have specific lint rules, probably)

**Comment by naugtur** [comment]
Oh, so my autoformatter did a good thing. yay!

**Comment by HenrikJoreteg** [comment]
@latentflip was there something that wasn't 4 spaces or something? Anyway, I'm +1

**Comment by HenrikJoreteg** [comment]
@latentflip @naugtur not sure exactly where this left off or what's left, here. Does this all work happily in node and with both browserify and webpack now?

Also, @naugtur you're clearly very meticulous about details so I've also given you publishing rights on npm for this module. Here's a bit about how we've been doing releases: http://ampersandjs.com/learn/bug-triage-process/#publishing at the point where any of the PRs you're involved in are ready per the guidelines on that page, fire away.

**Comment by naugtur** [comment]
Thanks for the access, I'll do the reading.
This change was supposed to be one of the main items in v4 of sync, so I'd rather not do my first publishing on a major version update ;) 

I made the testrunner run everything on node also. Didn't test cross-browser though. 
I hoped someone would gather all the info about changes planned for v4 and decide to merge. That's why I wait.

I don't feel like shipping the v4 myself, so @latentflip, I hope you'll have time to finalize that... with my help if needed ;)

**Comment by naugtur** [comment]
Ok, I'm convinced ;)

**Comment by HenrikJoreteg** [comment]
hey @naugtur, so is this just waiting for someone to put this on npm? I gave you access. I believe in "open" open source :) Ship it!

http://ampersandjs.com/learn/bug-triage-process#publishing

**Comment by naugtur** [comment]
I know that, and I was meaning to say I'll do it by my previous comment here. 
This change was supposed to be released as 4.0 so before I release it, everything planned for 4.0 has to go in. I fixed 2 more things, merged in some stuff and there's one more thing left to implement before I can release 4.0

**Pull request merged**